components/proxy/src/main/kotlin/com/hotels/styx/services/YamlFileConfigurationService.kt

```diff
@@ -0,0 +1,146 @@
+import com.hotels.styx.routing.db.StyxObjectStore
+import com.hotels.styx.routing.handlers.ProviderObjectRecord
+import com.hotels.styx.serviceproviders.ServiceProviderFactory
+import com.hotels.styx.services.OriginsConfigConverter.Companion.deserialiseOrigins
+import org.slf4j.LoggerFactory
+import java.util.concurrent.atomic.AtomicReference
+
+internal class YamlFileConfigurationService(
+        private val routeDb: StyxObjectStore<RoutingObjectRecord>,
+        private val converter: OriginsConfigConverter,
+        private val config: YamlFileConfigurationServiceConfig,
+        val serviceDb: StyxObjectStore<ProviderObjectRecord>) : StyxService {
+
+    val fileMonitoringService = FileMonitoringService("YamlFileCoinfigurationService", config.originsFile) {
+        reloadAction(it)
+    }
+
+    val routingObjects = AtomicReference<List<Pair<String, RoutingObjectRecord>>>(listOf())
+    val healthMonitors = AtomicReference<List<Pair<String, ProviderObjectRecord>>>(listOf())
+
+    companion object {
+        @JvmField
+        val SCHEMA = SchemaDsl.`object`(
+                field("originsFile", string()),
+                optional("monitor", bool()))
+
+        private val LOGGER = LoggerFactory.getLogger(YamlFileConfigurationService::class.java)
+    }
+
+    override fun start() = fileMonitoringService.start()
+            .thenAccept {
+                LOGGER.info("service started - {} - {}", config.originsFile)
+            }
+
+    override fun stop() = fileMonitoringService.stop()
+            .thenAccept {
+                LOGGER.info("service stopped")
+            }
+
+    fun reloadAction(content: String): Unit {
+        LOGGER.info("New origins configuration: \n$content")
+
+        kotlin.runCatching {
+            val deserialised = deserialiseOrigins(content)
+
+            val routingObjects = converter.routingObjects(deserialised)
+            val healthMonitors = converter.healthCheckServices(deserialised)
+
+            Pair(healthMonitors, routingObjects)
+        }.mapCatching { (healthMonitors, routingObjects) ->
+            updateRoutingObjects(routingObjects)
+            updateHealthCheckServices(serviceDb, healthMonitors)
+        }.onFailure {
+            LOGGER.error("Failed to reload new configuration. cause='{}'", it.message, it)
+        }
+    }
+
+    private fun changed(one: JsonNode, another: JsonNode) = !one.equals(another)
+
+    internal fun updateRoutingObjects(objects: List<Pair<String, RoutingObjectRecord>>) {
+        val oldObjectNames = routingObjects.get().map { it.first }
+        routingObjects.set(objects)
+
+        val newObjectNames = objects.map { it.first }
+        val removedObjects = oldObjectNames.minus(newObjectNames)
+
+        objects.forEach { (name, new) ->
+            routeDb.compute(name) { previous ->
+                if (previous == null || changed(new.config, previous.config)) {
+                    previous?.routingObject?.stop()
```

Comment:
I think routingObject cannot be null.

Reply:
Yes that is correct. But Kotlin compiler doesn't know that. Therefore I need use a safe call operator in `...routingObject?.stop()`.

Reply:
That's fair enough... I was not thinking what it would take to remove the second null check; probably something like  previous?.let{it.stop} which is way less readable.

Reply:
Okay. I'll resolve this conversation then :-)